ibis/expr/rules.py

```diff
@@ -224,6 +224,17 @@ def column(inner, arg):
     return instance_of(ir.ColumnExpr, inner(arg))
 
 
+@validator
+def array_of(inner, arg):
+    val = arg if isinstance(arg, ir.Expr) else ir.literal(arg)
+    argtype = val.type()
+    if not isinstance(argtype, dt.Array):
+        raise com.IbisTypeError(
+            'Argument must be an array, got expression {} which is of type '
+            '{}'.format(val, val.type()))
+    return value(dt.Array(inner(val[0]).type()), val)
```

Comment:
@cpcloud Is this for more verbose error? Because 

```python
value(dt.Array(dt.Array(dt.string)), [['a'], [], [], ['a', 1]])
``` 
would work.

Reply:
No, I want to be able to use rules for the element type.